toolkits/google/arcade_google/tools/gmail.py

```diff
@@ -506,16 +610,101 @@ async def get_thread(
+
     thread = service.users().threads().get(**params).execute()
-    thread["messages"] = [parse_email(message) for message in thread.get("messages", [])]
+    thread["messages"] = [parse_plain_text_email(message) for message in thread.get("messages", [])]
 
     return dict(thread)
+
+
+@tool(
+    requires_auth=Google(
+        scopes=["https://www.googleapis.com/auth/gmail.modify"],
+    )
+)
+async def change_email_labels(
+    context: ToolContext,
+    email_id: Annotated[str, "The ID of the email to modify labels for"],
+    labels_to_add: Annotated[list[str], "List of label names to add"],
+    labels_to_remove: Annotated[list[str], "List of label names to remove"],
+) -> Annotated[dict, "List of labels that were added, removed, and not found"]:
+    """
+    Add and remove labels from an email using the Gmail API.
+    """
+    service = _build_gmail_service(context)
+
+    add_labels = get_label_ids(service, labels_to_add)
+    remove_labels = get_label_ids(service, labels_to_remove)
+
+    invalid_labels = (
+        set(labels_to_add + labels_to_remove) - set(add_labels.keys()) - set(remove_labels.keys())
+    )
+
+    if invalid_labels:
+        # prepare the list of valid labels
+        labels = service.users().labels().list(userId="me").execute().get("labels", [])
+        label_names = [label["name"] for label in labels]
+
+        # raise a retryable error with the list of valid labels
+        raise RetryableToolError(
+            message=f"Invalid labels: {invalid_labels}",
+            developer_message=f"Invalid labels: {invalid_labels}",
+            additional_prompt_content=f"List of valid labels: {label_names}",
+        )
+
+    # Prepare the modification body with label IDs.
+    body = {
+        "addLabelIds": list(add_labels.values()),
+        "removeLabelIds": list(remove_labels.values()),
+    }
+
+    try:  # Modify the email labels.
+        service.users().messages().modify(userId="me", id=email_id, body=body).execute()
+
+    except Exception as e:
+        raise GmailToolError(
+            message=f"Error modifying labels for email {email_id}", developer_message=str(e)
+        )
+
+    # Confirmation JSON with lists for added and removed labels.
+    confirmation = {
+        "addedLabels": list(add_labels.keys()),
+        "removedLabels": list(remove_labels.keys()),
+    }
+
+    return {"confirmation": dict(confirmation)}
+
+
+@tool(
+    requires_auth=Google(
+        scopes=["https://www.googleapis.com/auth/gmail.readonly"],
+    )
+)
+async def list_labels(
+    context: ToolContext,
+) -> Annotated[dict, "A dictionary containing a list of label details"]:
+    """List all the labels in the user's mailbox."""
+
+    service = _build_gmail_service(context)
+
+    labels = service.users().labels().list(userId="me").execute().get("labels", [])
+
+    return {"labels": labels}
+
+
+@tool(
+    requires_auth=Google(
+        scopes=["https://www.googleapis.com/auth/gmail.readonly"],
```

Comment:
incorrect scope for this action. valid scopes listed here: https://developers.google.com/gmail/api/reference/rest/v1/users.labels/create#authorization-scopes